core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/NetworkTopologyReplicationStrategy.java

```diff
@@ -56,21 +56,31 @@
   }
 
   @Override
-  public SetMultimap<Token, Node> computeReplicasByToken(
+  public Map<Token, Set<Node>> computeReplicasByToken(
       Map<Token, Node> tokenToPrimary, List<Token> ring) {
 
-    // This is essentially a copy of org.apache.cassandra.locator.NetworkTopologyStrategy
-    ImmutableSetMultimap.Builder<Token, Node> result = ImmutableSetMultimap.builder();
+    // The implementation of this method was adapted from
+    // org.apache.cassandra.locator.NetworkTopologyStrategy
+
+    ImmutableMap.Builder<Token, Set<Node>> result = ImmutableMap.builder();
     Map<String, Set<String>> racks = getRacksInDcs(tokenToPrimary.values());
     Map<String, Integer> dcNodeCount = Maps.newHashMapWithExpectedSize(replicationFactors.size());
     Set<String> warnedDcs = Sets.newHashSetWithExpectedSize(replicationFactors.size());
+    // To optimize memory usage, ensure that identical replica sets will be represented by a single
+    // instance. Order matters so the key must be a list.
+    Map<List<Node>, Set<Node>> canonicalReplicaSets = new HashMap<>();
```

Comment:
Contrary to what I said on the ticket, we need a list as the key here. Otherwise sets in different order would get the same canonical representation (LinkedHashSet's hashCode and equals are not order-sensitive).

So I either had to convert `replicas` into a list when I do the lookup, or make it a list to begin with. I chose the latter to avoid yet another copy. This forces me to use the linear `addIfAbsent`, but the impact will be minimal since the cardinalities are low.